quickwit-metastore/src/metastore/single_file_metastore/mod.rs

```diff
@@ -0,0 +1,575 @@
+        // That's because we want to observe the property that after one success
+        // all subsequent calls will succeed.
+        let metadata_set_result = fetch_metadata_set(&*self.storage, index_id).await;
+
+        // Here we retake the lock, still no io ongoing.
+        let mut per_index_metastores_wlock = self.per_index_metastores.write().await;
+        if let Some(metadata_set) = per_index_metastores_wlock.get(index_id) {
+            return Ok(metadata_set.clone());
+        }
+
+        let metadata_set = metadata_set_result?;
+        let metadata_set_mutex = Arc::new(Mutex::new(metadata_set));
+        per_index_metastores_wlock.insert(index_id.to_string(), metadata_set_mutex.clone());
+        Ok(metadata_set_mutex)
+    }
+
+    // Helper used for testing to obtain the data associated with the given index.
+    #[cfg(test)]
+    async fn get_index(&self, index_id: &str) -> MetastoreResult<MetadataSet> {
+        self.read(index_id, |metadata_set| Ok(metadata_set.clone()))
+            .await
+    }
+
+    #[cfg(test)]
+    async fn index_exists(&self, index_id: &str) -> MetastoreResult<bool> {
+        index_exists(&*self.storage, index_id).await
+    }
+}
+
+#[async_trait]
+impl Metastore for SingleFileMetastore {
+    /// -------------------------------------------------------------------------------
+    /// Mutations over the high-level index.
+
+    async fn create_index(&self, index_metadata: IndexMetadata) -> MetastoreResult<()> {
+        // We pick the outer lock here, so that we enter a critical section.
+        let mut per_index_metastores_wlock = self.per_index_metastores.write().await;
+
+        let exists = per_index_metastores_wlock.contains_key(&index_metadata.index_id)
+            || index_exists(&*self.storage, &index_metadata.index_id).await?;
+
+        if exists {
+            return Err(MetastoreError::IndexAlreadyExists {
+                index_id: index_metadata.index_id.clone(),
+            });
+        }
+
+        let metadata_set = MetadataSet::from(index_metadata);
+
+        put_metadata_set(&*self.storage, &metadata_set).await?;
+
+        let index_id = metadata_set.index_id().to_string();
+        let metadata_set_mutex = Arc::new(Mutex::new(metadata_set));
+
+        per_index_metastores_wlock.insert(index_id, metadata_set_mutex);
+
+        Ok(())
+    }
+
+    async fn delete_index(&self, index_id: &str) -> MetastoreResult<()> {
+        // We pick the outer lock here, so that we enter a critical section.
+        let mut per_index_metastores_wlock = self.per_index_metastores.write().await;
+
+        let delete_res = delete_metadata_set(&*self.storage, index_id).await;
+
+        match &delete_res {
+            Ok(()) |
+            // If the index file does not exist, we still need to return an error,
+            // but it makes sense to ensure that the cache is up to date.
+            Err(MetastoreError::IndexDoesNotExist { .. }) => {
```

Comment:
we are guarding against the index file being deleted by an external process right?

Reply:
Technically we do not support two processes writing on the singlefilemetastore.

Still it is nice to have a proper definition of a trait like the `Metastore` on that kind of corner case that "should never happen".